pallets/ocex/src/lib.rs

```diff
@@ -734,45 +735,72 @@ pub mod pallet {
 			// Anyone can claim the withdrawal for any user
 			// This is to build services that can enable free withdrawals similar to CEXes.
 			let _ = ensure_signed(origin)?;
-
-			let mut withdrawals: WithdrawalsMap<T> = <Withdrawals<T>>::get(snapshot_id);
-			ensure!(withdrawals.contains_key(&account), Error::<T>::InvalidWithdrawalIndex);
-			if let Some(withdrawal_vector) = withdrawals.get(&account) {
-				for x in withdrawal_vector.iter() {
-					// TODO: Security: if this fails for a withdrawal in between the iteration, it
-					// will double spend.
-					if let Some(converted_withdrawal) =
-						x.amount.saturating_mul(Decimal::from(UNIT_BALANCE)).to_u128()
-					{
-						Self::transfer_asset(
-							&Self::get_custodian_account(),
-							&x.main_account,
-							converted_withdrawal.saturated_into(),
-							x.asset,
-						)?;
+			// This vector will keep track of withdrawals processed already
+			let mut processed_withdrawals: BoundedVec<Withdrawal<T::AccountId>, WithdrawalLimit> =
+				BoundedVec::<Withdrawal<T::AccountId>, WithdrawalLimit>::default();
+			ensure!(
+				<Withdrawals<T>>::contains_key(&snapshot_id),
+				Error::<T>::InvalidWithdrawalIndex
+			);
+			// This entire block of code is put inside ensure as some of the nested functions will
+			// return Err
+			ensure!(
+				<Withdrawals<T>>::mutate(snapshot_id, |btree_map| {
+					// Get mutable reference to the withdrawals vector
+					if let Some(withdrawal_vector) = btree_map.get_mut(&account) {
+						while withdrawal_vector.len() > 0 {
+							// Perform pop operation to ensure we do not leave any withdrawal left
+							// for a double spend
+							if let Some(withdrawal) = withdrawal_vector.pop() {
+								if let Some(converted_withdrawal) = withdrawal
+									.amount
+									.saturating_mul(Decimal::from(UNIT_BALANCE))
+									.to_u128()
+								{
+									if Self::transfer_asset(
+										&Self::get_custodian_account(),
+										&withdrawal.main_account,
+										converted_withdrawal.saturated_into(),
+										withdrawal.asset,
+									)
+									.is_ok()
+									{
+										processed_withdrawals
+											.try_push(withdrawal)
+											.unwrap_or_default();
+									}
+								}
+							}
+						}
+						btree_map.remove(&account);
```

Comment:
Keep the failed withdrawals for later inspection.

Reply:
Done ✅